### ISSUE TITLE
no-any not recognized when parameter typedef is off

### ISSUE THREAD

**Issue by iMarv** [Microsoft/vscode-tslint#286]
My `tslint.json` has the `no-any` rule activated, but restricts typedef only this way:

```json
"typedef": [
  true,
  "call-signature",
  "parameter",
  "member-variable-declaration",
  "variable-declaration",
  "property-declaration",
  "arrow-call-signature"
],
```

With this combination, implicit any declarations are not considered as arrows by the extension, but when running tslint on the command line.
This happens when using the given tslint configuration and running async jasmine tests like this:

```typescript
it('should do something', (done) => {
    something.doStuff().then(value => {
        expect(value).toBe(0);
        done();
    });
});
```

`done` implicitly has the type `any` but is not recognized by the extension.

When running this configuration from command line, the expected error is thrown.

**Comment by egamma** [comment]
Can you please share your complete tslint.json so that I can try to reproduce the problem.

**Comment by iMarv** [comment]
Of course
```json
{
  "rulesDirectory": [
    "node_modules/codelyzer"
  ],
  "jsRules": {
    "class-name": true,
    "comment-format": [
      true,
      "check-space"
    ],
    "indent": [
      true,
      "spaces"
    ],
    "no-duplicate-variable": true,
    "no-eval": true,
    "no-trailing-whitespace": true,
    "no-unsafe-finally": true,
    "one-line": [
      true,
      "check-open-brace",
      "check-whitespace"
    ],
    "quotemark": [
      true,
      "single"
    ],
    "semicolon": [
      true,
      "always"
    ],
    "triple-equals": [
      true,
      "allow-null-check"
    ],
    "variable-name": [
      true,
      "ban-keywords"
    ],
    "whitespace": [
      true,
      "check-branch",
      "check-decl",
      "check-operator",
      "check-separator",
      "check-type"
    ]
  },
  "rules": {
    "adjacent-overload-signatures": true,
    "member-access": false,
    "member-ordering": [
      true,
      {
        "order": [
          "public-static-field",
          "protected-static-field",
          "private-static-field",
          "public-instance-field",
          "protected-instance-field",
          "private-instance-field",
          "public-static-method",
          "protected-static-method",
          "private-static-method",
          "public-constructor",
          "protected-constructor",
          "private-constructor",
          "public-instance-method",
          "protected-instance-method",
          "private-instance-method"
        ]
      }
    ],
    "no-any": true,
    "no-inferrable-types": false,
    "no-internal-module": true,
    "no-var-requires": true,
    "only-arrow-functions": false,
    "prefer-for-of": true,
    "typedef": [
      true,
      "call-signature",
      "parameter",
      "member-variable-declaration",
      "variable-declaration",
      "property-declaration",
      "arrow-call-signature"
    ],
    "typedef-whitespace": [
      true,
      {
        "call-signature": "nospace",
        "index-signature": "nospace",
        "parameter": "nospace",
        "property-declaration": "nospace",
        "variable-declaration": "nospace"
      },
      {
        "call-signature": "space",
        "index-signature": "space",
        "parameter": "space",
        "property-declaration": "space",
        "variable-declaration": "space"
      }
    ],
    "ban": false,
    "curly": true,
    "forin": true,
    "label-position": true,
    "no-arg": true,
    "no-bitwise": true,
    "no-conditional-assignment": true,
    "no-console": [
      true,
      "debug",
      "info",
      "time",
      "timeEnd",
      "trace"
    ],
    "no-construct": true,
    "no-debugger": true,
    "no-duplicate-variable": true,
    "no-empty": false,
    "no-eval": true,
    "no-null-keyword": false,
    "no-shadowed-variable": true,
    "no-string-literal": true,
    "no-switch-case-fall-through": true,
    "no-unused-expression": true,
    "no-use-before-declare": true,
    "no-var-keyword": true,
    "radix": true,
    "switch-default": true,
    "triple-equals": true,
    "use-isnan": true,
    "eofline": true,
    "cyclomatic-complexity": [true, 14],
    "indent": [
      true,
      "spaces"
    ],
    "max-line-length": [
      true,
      120
    ],
    "no-require-imports": true,
    "no-trailing-whitespace": true,
    "object-literal-sort-keys": false,
    "trailing-comma": [
      true,
      {
        "multiline": false,
        "singleline": "never"
      }
    ],
    "align": [
      true,
      "parameters",
      "statements"
    ],
    "array-type": [true, "array"],
    "class-name": true,
    "comment-format": [
      true,
      "check-space"
    ],
    "interface-name": false,
    "jsdoc-format": true,
    "new-parens": true,
    "no-consecutive-blank-lines": false,
    "object-literal-key-quotes": [
      true,
      "as-needed"
    ],
    "one-line": [
      true,
      "check-open-brace",
      "check-catch",
      "check-else",
      "check-finally",
      "check-whitespace"
    ],
    "ordered-imports": false,
    "quotemark": [
      true,
      "single",
      "avoid-escape"
    ],
    "semicolon": [
      true,
      "always"
    ],
    "variable-name": [
      true,
      "check-format",
      "allow-leading-underscore",
      "ban-keywords"
    ],
    "whitespace": [
      true,
      "check-branch",
      "check-decl",
      "check-operator",
      "check-separator",
      "check-type",
      "check-module"
    ],


    "directive-selector": [true, "attribute", [], "camelCase"],
    "component-selector": [true, "element", [], "kebab-case"],

    "use-input-property-decorator": true,
    "use-output-property-decorator": true,
    "use-host-property-decorator": false,
    "no-attribute-parameter-decorator": true,
    "no-input-rename": true,
    "no-output-rename": true,
    "no-forward-ref": true,
    "use-life-cycle-interface": true,
    "use-pipe-transform-interface": true,
    "component-class-suffix": true,
    "directive-class-suffix": true,
    "templates-use-public": true,
    "no-access-missing-member": false,
    "invoke-injectable": true
  }
}
```

**Comment by egamma** [comment]
@iMarv I'm not able to reproduce the issue, pls see the screenshot below. Running tslint from the command line and the warnings in test.ts shown by vscode-tslint are the same.

I've used your tslint.json remove the codelyzer rules, since they seam to be unrelated to the reported issue.

This is the test.ts I've used

```ts
describe('A suite is just a function', function() {
    it('should do something', (done) => {
        something.doStuff().then(value => {
            expect(value).toBe(0);
            done();
        });
    });
  });

```
I've used both tslint 5.8 on the command line and in the tool. Can you verify that you are using the same version of tslint on the command line as in vscode-tslint.

Questions:
- what is the output of tslint shown on the command line?
- inside vscode what is the output shown in the Output panel at the bottom when you open the 'tslint' channel?


![image](https://user-images.githubusercontent.com/172399/32416060-4696d07c-c243-11e7-8a66-7c56170730cd.png)

**Comment by aluanhaddad** [comment]
@iMarv 
> My `tslint.json` has the `no-any` rule activated, but restricts `typedef `only this way:

I'm not sure I understand. The [`no-any`](https://palantir.github.io/tslint/rules/no-any/) rule is to flag code such as
```ts
function f(x: any) { }
```
and
```ts
(window as any).$ = $;
```
and so on.

In other words it flags _explicit_ uses of `any`.


> With this combination, implicit any declarations are not considered as arrows by the extension, but when running tslint on the command line.

The rule never flagged _implicit_ uses of the `any` type because that is (a) a separate concern and (b) handled by TypeScript itself via `noImplicitAny`.

> This happens when using the given tslint configuration and running async jasmine tests like this:
> ```ts
> it('should do something', (done) => {
>     something.doStuff().then(value => {
>         expect(value).toBe(0);
>         done();
>     });
> });
> ```
> `done` implicitly has the type `any` but is not recognized by the extension.

Assuming you have `@types/jasmine` installed, neither the linter nor the compiler will bat an eyelash at `done` because it is well typed. I think the tslint configuration you are actually missing is
```json
"typedef": [true, "arrow-parameter"]
```
While, it is indeed strange that `"arrow-parameter"` and `"parameter"` are discrete rules, it is a very, very bad idea to use _either_ rule.
Consider the declaration 
```ts
declare function takesCallback(callback: (data: any) => void): void;
```
and the use
```ts
takesCallback((data: Response) => {
  console.log(data.json());
});
```
The above call to `takesCallback` appears to be well typed but can easily lead to the runtime error
```scala
Uncaught TypeError: data.json is not a function
```
In other words, the type annotation `Response` on `data` acts as a kind of hidden type assertion. This lulls us into a false sense of security because `data` will appear to have all of the members of `Response` and they will be displayed by intellisense.

Note: I use `no-any`(tslint) and `noImplicitAny`(typescript) in all of my projects.

**Comment by egamma** [comment]
@aluanhaddad thanks for this in-depth analysis!

**Comment by iMarv** [comment]
Sorry for the late reaction, and thank you for the explanation @aluanhaddad . This really gives me a better understanding of how the linter works.
